### PR TITLE
netexec: fix packaging

### PR DESCRIPTION
**Pull request by thiagokokada**
## Description of changes

Continuation of https://github.com/NixOS/nixpkgs/pull/275531, fixing a bunch of packaging mistakes from that PR.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/development/python-modules/jsonsir/default.nix [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+  };
+
+  nativeBuildInputs = [
+    setuptools
+  ];
+
+  pythonImportsCheck = [ "jsonsir" ];
+
+  # No tests in PyPi release
+  doCheck = false;
```
Comment (thiagokokada) [inline pkgs/development/python-modules/jsonsir/default.nix:25]: There are tests in the GitHub repository, however it is not tagged so it seems the PyPi release is more reliable in this case.

**Comment by thiagokokada** [comment]
CC @vncsb for testing.

CC @Janik-Haag and @mweinelt since you pointed to me the issues in the packaging.

File: pkgs/development/python-modules/dploot/default.nix [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    lxml
+  ];
+
+  pythonImportsCheck = [ "dploot" ];
+
+  # No tests
+  doCheck = false;
+
+  meta = {
+    homepage = "https://github.com/zblurx/dploot";
```
Comment (mweinelt) [inline pkgs/development/python-modules/dploot/default.nix:38]:
```suggestion
    homepage = "https://github.com/zblurx/dploot";
    changelog = "https://github.com/zblurx/dploot/releases/tag/${version}";
```

File: pkgs/development/python-modules/dploot/default.nix [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    description = "DPAPI looting remotely in Python";
+    license = lib.licenses.mit;
+    maintainers = with lib.maintainers; [ ];
```
Comment (mweinelt) [inline pkgs/development/python-modules/dploot/default.nix:41]: All python libraries should be maintained by the one introducing them, which would be @vncsb, who requires them for netexec.

File: pkgs/development/python-modules/python-easyconfig/default.nix [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    repo = "easyconfig";
+    rev = "v${version}";
+    hash = "sha256-B5Wubw/t6kltwLQdeOrO6QYg4MzIk1MjFXGVMSMp+mA=";
+  };
+
+  # Fix tests, otherwise we get the following error:
+  # Failed: [pytest] section in setup.cfg files is no longer supported, change to [tool:pytest] instead.
+  postPatch = ''
+    substituteInPlace setup.cfg --replace '[pytest]' '[tool:pytest]'
+  '';
```
Comment (mweinelt) [inline pkgs/development/python-modules/python-easyconfig/default.nix:26]: Send this upstream.

File: pkgs/development/python-modules/jsonform/default.nix [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+  src = fetchPypi {
+    inherit version;
+    pname = "JsonForm";
+    hash = "sha256-cfi3ohU44wyphLad3gTwKYDNbNwhg6GKp8oC2VCZiOY=";
+  };
+
+  postPatch = ''
+    substituteInPlace jsonform/__init__.py \
+      --replace "from jsonform import JsonForm" "from .jsonform import JsonForm"
+  '';
```
Comment (mweinelt) [inline pkgs/development/python-modules/jsonform/default.nix:23]: Send this upstream.

File: pkgs/development/python-modules/python-easyconfig/default.nix [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+
+  nativeCheckInputs = [
+    pytestCheckHook
+  ];
+
+  disabledTests = [
+    # TypeError: load() missing 1 required positional argument: 'Loader'
+    "test_valid_yaml_mapping"
+    "test_valid_yaml_mapping_from_envvar"
+  ];
```
Comment (mweinelt) [inline pkgs/development/python-modules/python-easyconfig/default.nix:47]: This is due to a lack of compat with recent pyyaml versions. The library needs to be updated, likely to use `safe_load`. Please send the change upstream.

https://github.com/RussellLuo/easyconfig/blob/master/easyconfig/mappings.py#L40
https://pyyaml.org/wiki/PyYAMLDocumentation#loading-yaml

**Review by mweinelt — DISMISSED** [review]
Not a fan of introducing libraries, that upstream hasn't touched in 7–10 years, that also show lack of compat with recent Python and common libraries like pyyaml.

File: pkgs/development/python-modules/python-easyconfig/default.nix (hunk 4 above)
Comment (thiagokokada) [reply pkgs/development/python-modules/python-easyconfig/default.nix:26]: Done.

File: pkgs/development/python-modules/dploot/default.nix (hunk 2 above)
Comment (thiagokokada) [reply pkgs/development/python-modules/dploot/default.nix:38]: Done.

File: pkgs/development/python-modules/jsonform/default.nix (hunk 5 above)
Comment (thiagokokada) [reply pkgs/development/python-modules/jsonform/default.nix:23]: Done.

File: pkgs/development/python-modules/python-easyconfig/default.nix (hunk 6 above)
Comment (thiagokokada) [reply pkgs/development/python-modules/python-easyconfig/default.nix:47]: Done.

**Comment by thiagokokada** [comment]
CC @mweinelt.

**Comment by vncsb** [comment]
I could build `netexec` with no problems but when I was testing with nixpkgs-review `dploot` seems to fail when checking runtime dependencies:

```
[nix-shell:~/.cache/nixpkgs-review/pr-279879-1]$ nix-store -l /nix/store/amvkxl0a458zprxc2g0jcfnbsjd0ab0a-python3.10-dploot-2.2.4.drv
Sourcing python-remove-tests-dir-hook
Sourcing python-catch-conflicts-hook.sh
Sourcing python-remove-bin-bytecode-hook.sh
Sourcing pypa-build-hook
Using pypaBuildPhase
Sourcing python-runtime-deps-check-hook
Using pythonRuntimeDepsCheckHook
Sourcing pypa-install-hook
Using pypaInstallPhase
Sourcing python-imports-check-hook.sh
Using pythonImportsCheckPhase
Sourcing python-namespaces-hook
Sourcing python-catch-conflicts-hook.sh
@nix { "action": "setPhase", "phase": "unpackPhase" }
Running phase: unpackPhase
unpacking source archive /nix/store/4k2qs1p3s6azlvkn0lknkzsidxrp3j8w-dploot-2.2.4.tar.gz
source root is dploot-2.2.4
setting SOURCE_DATE_EPOCH to timestamp 1701366012 of file dploot-2.2.4/pyproject.toml
@nix { "action": "setPhase", "phase": "patchPhase" }
Running phase: patchPhase
@nix { "action": "setPhase", "phase": "updateAutotoolsGnuConfigScriptsPhase" }
Running phase: updateAutotoolsGnuConfigScriptsPhase
@nix { "action": "setPhase", "phase": "configurePhase" }
Running phase: configurePhase
no configure script, doing nothing
@nix { "action": "setPhase", "phase": "buildPhase" }
Running phase: buildPhase
Executing pypaBuildPhase
Creating a wheel...
* Getting build dependencies for wheel...
* Building wheel...
Successfully built dploot-2.2.4-py3-none-any.whl
Finished creating a wheel...
Finished executing pypaBuildPhase
@nix { "action": "setPhase", "phase": "pythonRuntimeDepsCheckHook" }
Running phase: pythonRuntimeDepsCheckHook
Executing pythonRuntimeDepsCheck
Checking runtime dependencies for dploot-2.2.4-py3-none-any.whl
  - cryptography<41.0.0,>=40.0.1 not satisfied by version 41.0.7
  - lxml==4.9.3 not satisfied by version 4.9.4
  - pyasn1<0.5.0,>=0.4.8 not satisfied by version 0.5.1
```

Don't know if this would only happen in `nixpkgs-review` but I think relaxing deps would resolve this?

**Comment by mweinelt** [comment]
Your testing and running nixpkgs-review likely were before and after the introduction of the pythonRuntimeDepsCheck, so the issue is real and needs to be addressed. Conveniently the recommendation to use `pythonRelaxDepsHook` was cut from the log.

**Comment by thiagokokada** [comment]
> Don't know if this would only happen in nixpkgs-review but I think relaxing deps would resolve this?

Seems to be an issue only when building `dploot` with Python 3.11, but since the default still seems to be Python 3.10 it worked because `dploot` is being build as application.

**Comment by thiagokokada** [comment]
It seems there is some changes in `master`, and this is why the branch was working fine but `nixpkgs-review` doesn't (since `nixpkgs-review` merges `master` in the branch before building).

I added `pythonRelaxDepsHook` in `dploot` and now its build is working again. However now I am now getting a strange issue while building `netexec` itself:

```
       last 10 log lines:
       > adding 'pywerview-0.6.dist-info/top_level.txt'
       > adding 'pywerview-0.6.dist-info/RECORD'
       > removing build/bdist.linux-x86_64/wheel
       > Successfully built pywerview-0.6-py3-none-any.whl
       > Finished creating a wheel...
       > Finished executing pypaBuildPhase
       > Running phase: pythonRuntimeDepsCheckHook
       > Executing pythonRuntimeDepsCheck
       > Checking runtime dependencies for pywerview-0.6-py3-none-any.whl
       >   - impacket>=0.9.22 not satisfied by version 0.12.0.dev1
`` `

For some reason `pywerview` can't be build with the custom `impacket` even if it technically fits its version requirements. I even tried to add `pythonRelaxDepsHook` inside `pywerview` override in `netexec`, but the error continues.

**Comment by thiagokokada** [comment]
> For some reason pywerview can't be build with the custom impacket even if it technically fits its version requirements. I even tried to add pythonRelaxDepsHook inside pywerview override in netexec, but the error continues.

I changed the version metadata in the custom `impacket` from `0.12.0.dev1` to `0.12.0` and this seems to have do the trick.

**Comment by thiagokokada** [comment]
I fixed one last issue in `asn1tools` that also seems to be from recent changes in `master`.

And also did some additional changes in the `netexec` package, and it should build successfully.

**Comment by thiagokokada** [comment]
CC @mweinelt @Janik-Haag for review.

CC @vncsb for testing.

**Comment by thiagokokada** [comment]
Rebased to fix merge conflicts with `master`.

**Comment by thiagokokada** [comment]
> Not a fan of introducing libraries, that upstream hasn't touched in 7–10 years, that also show lack of compat with recent Python and common libraries like pyyaml.

Looking at the https://github.com/Pennyw0rth/NetExec/commit/483dc69a2a7aa8f364adfc46096a8b5114c0a31a when upstream introduced the dependency in `resource`, using https://github.com/RussellLuo/resource was a clear mistake from upstream. They probably were supposed to include https://docs.python.org/3/library/resource.html, that is in stdlib instead.

**Comment by thiagokokada** [comment]
> > Not a fan of introducing libraries, that upstream hasn't touched in 7–10 years, that also show lack of compat with recent Python and common libraries like pyyaml.
> 
> Looking at the [Pennyw0rth/NetExec@483dc69](https://github.com/Pennyw0rth/NetExec/commit/483dc69a2a7aa8f364adfc46096a8b5114c0a31a) when upstream introduced the dependency in `resource`, using https://github.com/RussellLuo/resource was a clear mistake from upstream. They probably were supposed to include https://docs.python.org/3/library/resource.html, that is in stdlib instead.
> 
> I just removed https://github.com/RussellLuo/resource, and this removed all those unmaintained dependencies.

BTW, I opened a PR upstream (https://github.com/Pennyw0rth/NetExec/pull/162), but since it doesn't apply cleanly in the current release I will leave the current approach of just removing the deps.

CC @mweinelt, with this change I think all your concerns in this PR should be fixed. Can you re-review?

**Comment by thiagokokada** [comment]
> > > Not a fan of introducing libraries, that upstream hasn't touched in 7–10 years, that also show lack of compat with recent Python and common libraries like pyyaml.
> > 
> > 
> > Looking at the [Pennyw0rth/NetExec@483dc69](https://github.com/Pennyw0rth/NetExec/commit/483dc69a2a7aa8f364adfc46096a8b5114c0a31a) when upstream introduced the dependency in `resource`, using https://github.com/RussellLuo/resource was a clear mistake from upstream. They probably were supposed to include https://docs.python.org/3/library/resource.html, that is in stdlib instead.
> > I just removed https://github.com/RussellLuo/resource, and this removed all those unmaintained dependencies.
> 
> BTW, I opened a PR upstream ([Pennyw0rth/NetExec#162](https://github.com/Pennyw0rth/NetExec/pull/162)), but since it doesn't apply cleanly in the current release I will leave the current approach of just removing the deps.
> 
> CC @mweinelt, with this change I think all your concerns in this PR should be fixed. Can you re-review?

Upstream merged my patch, I just switched this to unstable for now.

**Comment by exploitoverload** [comment]
Hi @thiagokokada !
I just ran a nixpkgs-review and got two errors. One in nix build and the other when running the tool itself. The error in nix build is found when compiling crackmapexec.

```
error: builder for '/nix/store/j72ppqdkbaa7ny5zvln69yv9vbvi06n9-crackmapexec-5.4.0.drv' failed with exit code 1;
       last 10 log lines:
       > Finished creating a wheel...
       > /build/source/dist /build/source
       > Unpacking to: unpacked/crackmapexec-5.4.0...OK
       > Repacking wheel as ./crackmapexec-5.4.0-py3-none-any.whl...OK
       > /build/source
       > Finished executing pypaBuildPhase
       > Running phase: pythonRuntimeDepsCheckHook
       > Executing pythonRuntimeDepsCheck
       > Checking runtime dependencies for crackmapexec-5.4.0-py3-none-any.whl
       >   - neo4j not satisfied by version 5.16.dev0
       For full logs, run 'nix log /nix/store/j72ppqdkbaa7ny5zvln69yv9vbvi06n9-crackmapexec-5.4.0.drv'.
error: 1 dependencies of derivation '/nix/store/7b06w3ldjcnqny98v6zbzfmwlr3pzf4q-review-shell.drv' failed to build
```
It seems to be a similar dependency error as dploot or pywerview, this time with neo4j.

The other error is when running nxc with smb or ssh does not find the module impacket.dcerpc.v5.gkdi:

```
Traceback (most recent call last):
  File "/nix/store/66fi4py61wqr9zxhbvb94qjfmb7q6caw-netexec-1.1.0-unstable-2024-01-15/bin/.nxc-wrapped", line 9, in <module>
    sys.exit(main())
             ^^^^^^
  File "/nix/store/66fi4py61wqr9zxhbvb94qjfmb7q6caw-netexec-1.1.0-unstable-2024-01-15/lib/python3.11/site-packages/nxc/netexec.py", line 152, in main
    protocol_object = getattr(p_loader.load_protocol(protocol_path), args.protocol)
                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/nix/store/66fi4py61wqr9zxhbvb94qjfmb7q6caw-netexec-1.1.0-unstable-2024-01-15/lib/python3.11/site-packages/nxc/loaders/protocolloader.py", line 16, in load_protocol
    loader.exec_module(protocol)
  File "<frozen importlib._bootstrap_external>", line 940, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/nix/store/66fi4py61wqr9zxhbvb94qjfmb7q6caw-netexec-1.1.0-unstable-2024-01-15/lib/python3.11/site-packages/nxc/protocols/ssh.py", line 9, in <module>
    from nxc.connection import connection, highlight
  File "/nix/store/66fi4py61wqr9zxhbvb94qjfmb7q6caw-netexec-1.1.0-unstable-2024-01-15/lib/python3.11/site-packages/nxc/connection.py", line 13, in <module>
    from nxc.protocols.ldap.laps import laps_search
  File "/nix/store/66fi4py61wqr9zxhbvb94qjfmb7q6caw-netexec-1.1.0-unstable-2024-01-15/lib/python3.11/site-packages/nxc/protocols/ldap/laps.py", line 11, in <module>
    from impacket.dcerpc.v5.gkdi import MSRPC_UUID_GKDI, GkdiGetKey, GroupKeyEnvelope
ModuleNotFoundError: No module named 'impacket.dcerpc.v5.gkdi'
```

In this other case, the truth is that I don't know why it doesn't detect the module if it is present in impacket.

**Comment by thiagokokada** [comment]
> I just ran a nixpkgs-review and got two errors. One in nix build and the other when running the tool itself. The error in nix build is found when compiling crackmapexec.

Not really related to this PR, I assume `crackmapexec` is broken in `master` already?

**Comment by thiagokokada** [comment]
Result of `nixpkgs-review` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages marked as broken and skipped:</summary>
  <ul>
    <li>python312Packages.dploot</li>
    <li>python312Packages.dploot.dist</li>
  </ul>
</details>
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>netexec</li>
    <li>netexec.dist</li>
    <li>python311Packages.dploot</li>
    <li>python311Packages.dploot.dist</li>
  </ul>
</details>

**Comment by thiagokokada** [comment]
Hey @mweinelt, can I get your review here? I think the package now is in a good state to merge, and all your original issues should have been fixed.

**Comment by thiagokokada** [comment]
CC @exploitoverload @vncsb for testing.

File: pkgs/tools/security/netexec/default.nix [hunk 7]
```diff
@@ -1,172 +1,105 @@
...
 
   meta = with lib; {
-    description = "Network service exploitation tool (Maintaned fork of CrackMapExec)";
+    description = "Network service exploitation tool (maintained fork of CrackMapExec)";
     homepage = "https://github.com/Pennyw0rth/NetExec";
     changelog = "https://github.com/Pennyw0rth/NetExec/releases/tag/v${version}";
     license = with licenses; [ bsd2 ];
     mainProgram = "nxc";
     maintainers = with maintainers; [ vncsb ];
+    broken = stdenv.isDarwin;
```
Comment (SuperSandro2000) [inline pkgs/tools/security/netexec/default.nix:103]: We should add a short comment with the reason

File: pkgs/tools/security/netexec/default.nix [hunk 8]
```diff
@@ -1,172 +1,105 @@
...
-    hash = "sha256-cNkZoIdfrKs5ZvHGKGBybCWGwA6C4rqjCOEM+pX70S8=";
+    rev = "9df72e2f68b914dfdbd75b095dd8f577e992615f";
+    hash = "sha256-oQHtTE5hdlxHX4uc412VfNUrN0UHVbwI0Mm9kmJpNW4=";
   };
 
+  postPatch = ''
+    substituteInPlace pyproject.toml \
+      --replace '{ git = "https://github.com/Pennyw0rth/impacket.git", branch = "gkdi" }' '"*"'
+
+    substituteInPlace pyproject.toml \
```
Comment (SuperSandro2000) [inline pkgs/tools/security/netexec/default.nix:47]:
```suggestion
      --replace '{ git = "https://github.com/Pennyw0rth/impacket.git", branch = "gkdi" }' '"*"' \
```

Comment (thiagokokada) [reply pkgs/tools/security/netexec/default.nix]: Done.

File: pkgs/tools/security/netexec/default.nix (hunk 7 above)
Comment (thiagokokada) [reply pkgs/tools/security/netexec/default.nix:104]: Done.

**Comment by AkechiShiro** [comment]
@SuperSandro2000 were your review comments all addressed, is there anything else that needs to be done to move forward this PR (besides testing ?)

**Pull request merged**